Qualify IPv6 link-local addresses with scope_id

ibygrave opened the pull request:
This replaces https://github.com/jstasiak/python-zeroconf/pull/315 which has been abandoned by the original author.
There are some outstanding problems with the original PR that I'll try to address:
* Is there a better way to provide this functionality than with a new separate API?
* Missing scope_id when IPv6 advertisement received over IPv4. https://github.com/jstasiak/python-zeroconf/pull/315#issuecomment-809827529
* Test coverage decrease.
* Travis CI failed on pypy3.

Closes #314

ibygrave commented:
Github hosted test coverage of any IPv6 feature is dependent on https://github.com/actions/virtual-environments/issues/668

Local builds on python 3.8 with IPV6 tests enabled shows the test coverage increasing slightly.

jstasiak commented:
No, the test coverage is not an issue, we understand the IPv6 situation. I need to think about what I'd like this API to be, I'm thinking if we're introducing something new we might as well use the `IPv4Address` and `IPv6Address` classes from the `ipaddress` module instead of just strings, I'll get back to you on this.

bdraco commented:
@ibygrave The CI is now running IPv6 tests. Can you rebase this on `upstream/master`?

bdraco commented:
Thanks.  Sorry about the conflicts.

bdraco commented on zeroconf/_core.py:
```diff
@@ -571,7 +577,13 @@ def send(self, out: DNSOutgoing, addr: Optional[str] = None, port: int = _MDNS_P
         assert self.loop is not None
         self.loop.call_soon_threadsafe(self.async_send, out, addr, port)
```
We should probably make the threadsafe version support v6 as well

bdraco commented on zeroconf/_core.py:
```diff
@@ -198,12 +198,16 @@ def datagram_received(
+            _scope = None
         elif len(addrs) == 4:
             # https://github.com/python/mypy/issues/1178
             addr, port, _flow, _scope = addrs  # type: ignore
+            log.debug('IPv6 scope_id %d associated to the receiving interface', _scope)
+            _v6 = (_flow, _scope)
```
```suggestion
            addr, port, flow, scope = addrs  # type: ignore
            log.debug('IPv6 scope_id %d associated to the receiving interface', _scope)
            _v6 = (flow, scope)
```

These were only prefixed with underscores to indicate they were unused

bdraco commented on zeroconf/_core.py:
```diff
@@ -198,12 +198,16 @@ def datagram_received(
         if len(addrs) == 2:
             # https://github.com/python/mypy/issues/1178
             addr, port = addrs  # type: ignore
+            _scope = None
```
```suggestion
            scope = None
```

bdraco commented on zeroconf/_core.py:
```diff
@@ -219,7 +223,7 @@ def datagram_received(
 
         self.data = data
-        msg = DNSIncoming(data)
+        msg = DNSIncoming(data, _scope)
```
```suggestion
        msg = DNSIncoming(data, scope)
```

bdraco commented on zeroconf/_core.py:
```diff
@@ -198,12 +198,16 @@ def datagram_received(
         self, data: bytes, addrs: Union[Tuple[str, int], Tuple[str, int, int, int]]
     ) -> None:
         assert self.transport is not None
+        _v6: Union[Tuple[()], Tuple[int, int]] = ()
```
Why prefix with `_`?

ibygrave replied:
I'm not the original author, but my guess is cargo cult. I'll remove, and adjust the other names as you suggest.

ibygrave replied:
Trying it I see that pylint doesn't like `v6` as it's too short.

bdraco replied:
`v6_scope_flow` seems like a name that would show intent

bdraco commented on zeroconf/_dns.py:
```diff
@@ -226,10 +226,18 @@ class DNSAddress(DNSRecord):
     ) -> None:
         super().__init__(name, type_, class_, ttl, created)
         self.address = address
+        self.scope_id = scope_id
```
Does `__eq__` and `__hash__` need to be updated?

If the last record in the cache with the scope_id is the one that should always win then they don't since we will replace any existing records in the cache that match the address with the new scope id.

Since `ServiceInfo` gained the `interface_index` it seems like they don't need to change since the expectation is there can only be one.

ibygrave replied:
I think an interface index can change, if the interface is removed and then re-added. For example a USB ethernet adapter. You might only see this on Windows because interface indexes seems to be less deterministic there.

Anyway, implementing these dunder methods should be harmless, so I will.

bdraco reviewed (commented):
Thanks for this.  Some small comments above!

bdraco commented:
Apologizes for the relocations causing conflicts. They are completed now so there shouldn't be any more.

bdraco commented on zeroconf/_dns.py:
```diff
@@ -225,10 +225,18 @@ class DNSAddress(DNSRecord):
+        class_: int,
+        ttl: int,
+        address: bytes,
+        created: Optional[float] = None,
```
We should move created to the end to be consistent with the other records. It's now and hasn't shipped yet so it won't be a breaking change

ibygrave replied:
Should `scope_id` and `created` be made keyword-only arguments?

bdraco replied:
Yes, I think that makes sense.

bdraco commented on examples/browser.py:
```diff
@@ -21,8 +21,9 @@ def on_service_state_change(
+
         if info:
-            addresses = ["%s:%d" % (addr, cast(int, info.port)) for addr in info.parsed_addresses()]
+            addresses = ["%s:%d" % (addr, cast(int, info.port)) for addr in info.parsed_scoped_addresses()]
```
Can you update the async version as well?

bdraco reviewed (commented):
Looks good!  Some small comments above. Let me know when you have finished testing and you think it's ready to merge.   I'll give it a spin later today as well

ibygrave commented:
> Looks good! Some small comments above. Let me know when you have finished testing and you think it's ready to merge. I'll give it a spin later today as well

I was almost done, except for the keyword-only arguments. Integration testing failed with intermittent discovery failure. Using the `ServiceBrowser` API with a listener class, with 4 services to be discovered, on almost every test one of the services is not discovered while the rest are. I see the same failure using master at commit ecad4e84. Integration testing of this branch did pass where service discovery succeeded.

bdraco commented:
Do you have a test case for the failure?

Also if you clear the cache before running the test, does it solve the failure case?

bdraco commented:
@ibygrave I fixed a thread safety issue that could result in lost record to ServiceBrowsers in the startup window.  Let me know if you can still replicate the issue on `upstream/master`.

Thanks 👍

ibygrave commented:
> @ibygrave I fixed a thread safety issue that could result in lost record to ServiceBrowsers in the startup window. Let me know if you can still replicate the issue on `upstream/master`.

It looks like you've fixed the issue I was seeing. Thanks.

I've rebased and squashed this branch and retested. It is ready to merge.

bdraco commented on zeroconf/_dns.py:
```diff
@@ -257,12 +266,15 @@ def write(self, out: 'DNSOutgoing') -> None:
         return (
-            isinstance(other, DNSAddress) and DNSEntry.__eq__(self, other) and self.address == other.address
+            isinstance(other, DNSAddress)
+            and DNSEntry.__eq__(self, other)
```
It would be faster to reject mismatched addresses sooner then checking eq.  I Will adjust in a follow up

bdraco replied:
addressed here https://github.com/jstasiak/python-zeroconf/pull/808

bdraco reviewed (approved):
Looks good.  I'll do some testing later today

bdraco commented:
Testing looks good

<img width="821" alt="Screen Shot 2021-06-22 at 6 58 08 AM" src="https://user-images.githubusercontent.com/663432/122968418-3b7a7b00-d327-11eb-80e6-dd270411257d.png">

No significant change in performance

<img width="700" alt="Screen Shot 2021-06-22 at 6 59 20 AM" src="https://user-images.githubusercontent.com/663432/122968596-6a90ec80-d327-11eb-9992-6ed6e9b441f3.png">

bdraco commented:
Thanks @ibygrave
